### ISSUE TITLE
Make esc and outsideclick trigger CancelEvent

### ISSUE THREAD

**Issue by KasparScherrer** [vaadin/vaadin-crud-flow#203]
The Dialog of the CrudEditor is closable by esc and outsideClick, but a CancelEvent is not fired, therefore the cruds cancelListeners are never executed in that case.

But there is no way to get to the Dialog instance that wraps the editor view, so I have no options to make a workaround for this. I can not set the Dialog to not close on esc or outsideClick, and I also can not add a listener myself as it was discussed in [this forum thread](https://vaadin.com/forum/thread/18196258/18196512), because `Component::addListener` is protected.

This is important to me, because I run some preparation code before opening a new item, and if the user then cancels I need to run another piece of code that reverts the preparations. When the user cancels using the cancel button, everything works fine, but as of now, I have no way of fixing the issue when the user closes the editor with any of the other ways.

possible ways that would help me are:
- add an API to the `Crud` or `CrudEditor` to get the `Dialog` instance. 
```
crud.getEditor().getDialog().setCloseOnEsc(false);
```
- or add an API to define a `DialogCloseActionListener`, similar to `Dialog::addDialogCloseActionListener`
```
crud.addEditorDialogCloseAcionListener(event -> fireEvent(new Crud.CancelEvent(crud, false, false)));
```
- or make outsideClicks and esc also trigger CancelEvent by default. I'm not sure if there is any use-case where this shouldn't happen automatically.

**Comment by KasparScherrer** [comment]
I made a false assumption above that I couldn't listen to the `DialogCloseActionEvent` - I could listen to it from my view. But it seems like there is never such an event fired in the first place.

After closer investigation, I figured that since the dialog of the editor is only defined in the template but the crud-flow component holds no instance of `Dialog`, there is no code like `Dialog::doEnsureOnCloseConfigured` that will actually dispatch the event "vaadin-dialog-close-action" or listen to the events "vaadin-overlay-outside-click" and "vaadin-overlay-escape-press".

I now realize how complex this whole topic of opening and closing of Dialogs really is, and I have given up trying to find a workaround myself. 

However, I see this as a showstopper for my application, because not triggering my cleanup code when cancelling a new item can lead to wrong configurations of my foobars, which I must prevent with all means.

**Comment by KasparScherrer** [comment]
I did find a way to work around this. I did it by extending `Crud` and executing some javascript that adds listeners to the overlay close events, and dispatches my selfdefined vaadin-crud-editor-close-action event. I took that code from Dialog::doEnsureOnCloseConfigured and adapted it.

As a workaround to this issue, the following Crud-extension can be used. As a fix for the long-term, I hope vaadins `Crud` class will adapt this code so my extension will no longer be necessary.

```
public class Crud<E> extends com.vaadin.flow.component.crud.Crud<E> {
    private int onCloseConfigured = 0;

    public Crud(Class<E> beanType, com.vaadin.flow.component.grid.Grid<E> grid, CrudEditor<E> editor) {
        super(beanType, grid, editor);
        makeCloseActionTriggerCancelEvent();
    }
    public Crud(Class<E> beanType, CrudEditor<E> editor) {
        super(beanType, editor);
        makeCloseActionTriggerCancelEvent();
    }
    public Crud() {
        super();
        makeCloseActionTriggerCancelEvent();
    }

    private void makeCloseActionTriggerCancelEvent() {
        ensureOnCloseConfigured();

        addListener(CrudEditorCloseActionEvent.class, e -> {
            fireEvent(new Crud.CancelEvent<>(this, false, false));
        });
    }

    private void ensureOnCloseConfigured() {
        if (onCloseConfigured == 0) {
            getElement().getNode()
                    .runWhenAttached(ui -> ui.beforeClientResponse(this,
                            context -> doEnsureOnCloseConfigured(ui)));
        }
        onCloseConfigured++;
    }

    private void doEnsureOnCloseConfigured(UI ui) {
        ui.getPage().executeJs(
                "var f = function(e) {"
                + "  if (e.type == 'vaadin-overlay-escape-press' && !$0.__isDirty ||"
                + "      e.type == 'vaadin-overlay-outside-click' && !$0.__isDirty) {"
                + "    e.preventDefault();"
                + "    $0.dispatchEvent(new CustomEvent('vaadin-crud-editor-close-action'));"
                + "  }" 
                + "};"
                + "$0.$.dialog.$.dialog.$.overlay.addEventListener('vaadin-overlay-outside-click', f);"
                + "$0.$.dialog.$.dialog.$.overlay.addEventListener('vaadin-overlay-escape-press', f);"
                , getElement());
    }
    
    /**
     * `vaadin-crud-editor-close-action` is sent when the user clicks outside the
     * editor overlay or presses the escape key.
     */
    @DomEvent("vaadin-crud-editor-close-action")
    public static class CrudEditorCloseActionEvent extends ComponentEvent<Crud> {
        public CrudEditorCloseActionEvent(Crud source, boolean fromClient) {
            super(source, fromClient);
        }
    }
}
```

**Comment by TatuLund** [comment]
I think the problem fundamentally originates to this. I.e. Crud relies on vaadin-confrim-dialog, and since it is not emitting the cancel event on e.g. esc, subsequently Crud inherits this issue as well: https://github.com/vaadin/vaadin-confirm-dialog/issues/41

**Comment by KasparScherrer** [comment]
No, the event 'vaadin-overlay-escape-press' is being fired just fine. My workaround here builds on that -> When that event is fired i dispatch my own event which is actually listened to. The problem is not that the event `vaadin-overlay-escape-press` is not fired, the problem is that it is not listened to by the Crud. Same with outside click

**Comment by TatuLund** [comment]
The fix #201 is included in version 1.1.8 of Enhanced CRUD  https://vaadin.com/directory/component/enhanced-crud

**Comment by MehdiHa96** [comment]
The easiest solution which I used and worked, was calling addNewListener, which will get triggered when New Item button is clicked, and then set the editor item to an empty one. Below is a sample:

crud (your own crud object) 
Object( the item you are filling your crud with)

crud.addNewListener( e -> crud.getEditor().setItem(new Object()) );


Hope this comes helpful to someone..
